mlir/lib/ExecutionEngine/LevelZeroRuntimeWrappers.cpp

```diff
@@ -0,0 +1,491 @@
+    for (auto pool : eventPools)
+      L0_SAFE_CALL(zeEventPoolDestroy(pool));
+  }
+
+  void createNewPool(size_t numEvents) {
+    ze_event_pool_desc_t eventPoolDesc = {};
+    eventPoolDesc.flags = ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
+    eventPoolDesc.count = numEvents;
+    eventPools.push_back(nullptr);
+    L0_SAFE_CALL(zeEventPoolCreate(rtCtx->context, &eventPoolDesc, 1,
+                                   &rtCtx->device, &eventPools.back()));
+    currentEventsLimit += numEvents;
+  }
+
+  ze_event_handle_t takeEvent() {
+    ze_event_handle_t event{nullptr};
+    if (availableEvents.size()) {
+      event = availableEvents.back();
+      availableEvents.pop_back();
+    } else {
+      if (currentEventsCnt == currentEventsLimit)
+        createNewPool(numEventsPerPool);
+      currentEventsCnt++;
+      ze_event_desc_t eventDesc = {
+          ZE_STRUCTURE_TYPE_EVENT_DESC, nullptr,
+          static_cast<uint32_t>(currentEventsCnt % numEventsPerPool),
+          ZE_EVENT_SCOPE_FLAG_DEVICE, ZE_EVENT_SCOPE_FLAG_HOST};
+      L0_SAFE_CALL(zeEventCreate(eventPools.back(), &eventDesc, &event));
+    }
+    takenEvents.insert(event);
+    return event;
+  }
+
+  void releaseEvent(ze_event_handle_t event) {
+    auto found = takenEvents.find(event);
+    assert(found != takenEvents.end());
+    takenEvents.erase(found);
+    L0_SAFE_CALL(zeEventHostReset(event));
+    availableEvents.push_back(event);
+  }
+};
+
+L0RtContext &getRtContext() {
+  thread_local static L0RtContext rtContext;
+  return rtContext;
+}
+
+DynamicEventPool &getDynamicEventPool() {
+  thread_local static DynamicEventPool dynEventPool{&getRtContext()};
+  return dynEventPool;
+}
+
+struct StreamWrapper {
+  // avoid event pointer invalidations
+  std::deque<ze_event_handle_t> implicitEventStack;
+  DynamicEventPool &dynEventPool;
+
+  StreamWrapper(DynamicEventPool &dynEventPool) : dynEventPool(dynEventPool) {}
+  ~StreamWrapper() { sync(); }
+
+  ze_event_handle_t *getLastImplicitEventPtr() {
+    // Assume current implicit events will not be used after `sync`.
+    return implicitEventStack.size() ? &implicitEventStack.back() : nullptr;
+  }
+
+  void sync(ze_event_handle_t explicitEvent = nullptr) {
+    ze_event_handle_t syncEvent{nullptr};
+    if (!explicitEvent) {
+      ze_event_handle_t *lastImplicitEventPtr = getLastImplicitEventPtr();
+      syncEvent = lastImplicitEventPtr ? *lastImplicitEventPtr : nullptr;
+    } else {
+      syncEvent = explicitEvent;
+    }
+    if (syncEvent)
+      L0_SAFE_CALL(zeEventHostSynchronize(
+          syncEvent, std::numeric_limits<uint64_t>::max()));
+    // All of the "implicit" events were signaled and are of no use, release
+    // them. "explicit" event must be "released" via mgpuEventDestroy
+    for (auto event : implicitEventStack)
+      dynEventPool.releaseEvent(event);
+    implicitEventStack.clear();
+  }
+
+  void enqueueOp(
+      std::function<void(ze_event_handle_t, uint32_t, ze_event_handle_t *)>
```

Comment:
`std::function` objects can be expensive to create, use `llvm::function_ref` or templated functor
```
template<typenamr F>
enqueueOp(F&& op)
```

Reply:
Thanks, done.